README.md
![RSMQ: Redis Simple Message Queue for Node.js](https://img.webmart.de/rsmq_wide.png)

# Redis Simple Message Queue
A lightweight message queue for Node.js that requires no dedicated queue server. Just a Redis server.

[![Build Status](https://secure.travis-ci.org/smrchy/rsmq.png?branch=master)](http://travis-ci.org/smrchy/rsmq)
[![Dependency Status](https://david-dm.org/smrchy/rsmq.svg)](https://david-dm.org/smrchy/rsmq)

**tl;dr:** If you run a Redis server and currently use Amazon SQS or a similar message queue you might as well use this fast little replacement. Using a shared Redis server multiple Node.js processes can send / receive messages.

## Features
* Lightweight: **Just Redis** and ~500 lines of javascript.
* Speed: Send/receive 10000+ messages per second on an average machine. It's **just Redis**.
* Guaranteed **delivery of a message to exactly one recipient** within a messages visibility timeout.
* Received messages that are not deleted will reappear after the visibility timeout.
* [Test coverage](http://travis-ci.org/smrchy/rsmq)
* A message is deleted by the message id. The message id is returned by the `sendMessage` and `receiveMessage` method.
* Messages stay in the queue unless deleted.
* Optional RESTful interface via [rest-rsmq](https://github.com/smrchy/rest-rsmq)
* Typescript Typings ❤️
* Optional Promise-based API (only if `Promise` is defined), just suffix your method with `Async`, eg: `sendMessage` -> `sendMessageAsync`, all queue methods are supported

**Note:** RSMQ uses the Redis EVAL command (LUA scripts) so the minimum Redis version is 2.6+. 

## Usage
* After creating a queue you can send messages to that queue.
* The messages will be handled in a **FIFO** (first in first out) manner unless specified with a delay.
* Every message has a unique `id` that you can use to delete the message. 
* The `sendMessage` method will return the `id` for a sent message.
* The `receiveMessage` method will return an `id` along with the message and some stats.
* Should you not delete the message it will be eligible to be received again after the visibility timeout is reached.
* Please have a look at the `createQueue` and `receiveMessage` methods described below for optional parameters like **visibility timeout** and **delay**.

## Installation
`npm install rsmq`

## Modules for RSMQ
To keep the core of *RSMQ* small additional functionality is available as modules:

* [**rsmq-worker**](https://github.com/mpneuried/rsmq-worker) Helper to implement a worker with RSMQ.
* [**rest-rsmq**](https://github.com/smrchy/rest-rsmq) A RESTful interface for RSMQ.
* [**rsmq-cli**](https://github.com/mpneuried/rsmq-cli) A command-line interface / Terminal client for RSMQ.
* [**rsmq-promise-native**](https://www.npmjs.com/package/rsmq-promise-native) Native promise interface for RSMQ.
* [**rsmq-promise**](https://www.npmjs.com/package/rsmq-promise) Promise interface for RSMQ (via Bluebird).

## RSMQ in other languages

The simplicity of RSMQ is useful in other languages. Here is a list of implementations in other languages:

* [**Rust**](https://github.com/Couragium/rsmq-async-rs) RSMQ Async for Rust. See [#107](https://github.com/smrchy/rsmq/issues/107)
* [**Java**](https://github.com/igr/jrsmq) RSMQ for Java. See [#48](https://github.com/smrchy/rsmq/issues/48)
* [**PHP**](https://github.com/michsindelar/PhpRSMQ) RSMQ for PHP (work in progress)
* [**CSharp/DotNet**](https://github.com/tontonrally/rsmqCsharp) RSMQ for .NET

Note: Should you plan to port RSQM to another language please make sure to have tests to ensure compatibility with all RSMQ clients. And of course: let me know so i can mention your port here.

## Methods

### Constructor
Creates a new instance of RSMQ.

Parameters:

* `host` (String): *optional (Default: "127.0.0.1")* The Redis server
* `port` (Number): *optional (Default: 6379)* The Redis port
* `options` (Object): *optional (Default: {})* The [Redis options](https://github.com/NodeRedis/node_redis#options-object-properties) object. 
* `client` (RedisClient): *optional* A existing redis client instance. `host` and `server` will be ignored.
* `ns` (String): *optional (Default: "rsmq")* The namespace prefix used for all keys created by RSMQ
* `realtime` (Boolean): *optional (Default: false)* Enable realtime PUBLISH of new messages (see the [Realtime section](#realtime))
* `password` (String): *optional (Default: null)* If your Redis server requires a password supply it here

Example:

```javascript
const RedisSMQ = require("rsmq");
const rsmq = new RedisSMQ( {host: "127.0.0.1", port: 6379, ns: "rsmq"} );
```

### Queue

#### createQueue(options, callback)
Create a new queue.

Parameters:

* `qname` (String): The Queue name. Maximum 160 characters; alphanumeric characters, hyphens (-), and underscores (_) are allowed.
* `vt` (Number): *optional* *(Default: 30)* The length of time, in seconds, that a message received from a queue will be invisible to other receiving components when they ask to receive messages. Allowed values: 0-9999999 (around 115 days)
* `delay` (Number): *optional* *(Default: 0)* The time in seconds that the delivery of all new messages in the queue will be delayed. Allowed values: 0-9999999 (around 115 days)
* `maxsize` (Number): *optional* *(Default: 65536)* The maximum message size in bytes. Allowed values: 1024-65536 and -1 (for unlimited size)

Returns:

* `1` (Number)

Example:

```javascript
rsmq.createQueue({ qname: "myqueue" }, function (err, resp) {
	if (err) {
		console.error(err)
		return
	}

	if (resp === 1) {
		console.log("queue created")
	}
});
```

#### listQueues(options, callback)
List all queues

Returns an array:

* `["qname1", "qname2"]`

Example:

```javascript
rsmq.listQueues(function (err, queues) {
	if (err) {
		console.error(err)
		return
	}

	console.log("Active queues: " + queues.join( "," ) )
});
```

#### deleteQueue(options, callback)
Deletes a queue and all messages.

Parameters:

* `qname` (String): The Queue name.

Returns:

* `1` (Number)

Example:

```javascript
rsmq.deleteQueue({ qname: "myqueue" }, function (err, resp) {
	if (err) {
		console.error(err)
		return
	}

	if (resp === 1) {
		console.log("Queue and all messages deleted.")
	} else {
		console.log("Queue not found.")
	}
});
```

#### getQueueAttributes(options, callback)
Get queue attributes, counter and stats

Parameters:

* `qname` (String): The Queue name.

Returns an object:

* `vt` (Number): The visibility timeout for the queue in seconds
* `delay` (Number): The delay for new messages in seconds
* `maxsize` (Number): The maximum size of a message in bytes
* `totalrecv` (Number): Total number of messages received from the queue
* `totalsent` (Number): Total number of messages sent to the queue
* `created` (Number): Timestamp (epoch in seconds) when the queue was created
* `modified` (Number): Timestamp (epoch in seconds) when the queue was last modified with `setQueueAttributes`
* `msgs` (Number): Current number of messages in the queue
* `hiddenmsgs` (Number): Current number of hidden / not visible messages. A message can be hidden while "in flight" due to a `vt` parameter or when sent with a `delay`

Example:

```javascript
rsmq.getQueueAttributes({ qname: "myqueue" }, function (err, resp) {
	if (err) {
		console.error(err);
		return;
	}

	console.log("==============================================");
	console.log("=================Queue Stats==================");
	console.log("==============================================");
	console.log("visibility timeout: ", resp.vt);
	console.log("delay for new messages: ", resp.delay);
	console.log("max size in bytes: ", resp.maxsize);
	console.log("total received messages: ", resp.totalrecv);
	console.log("total sent messages: ", resp.totalsent);
	console.log("created: ", resp.created);
	console.log("last modified: ", resp.modified);
	console.log("current n of messages: ", resp.msgs);
	console.log("hidden messages: ", resp.hiddenmsgs);
});
```

    
#### setQueueAttributes(options, callback)
Sets queue parameters.

Parameters:

* `qname` (String): The Queue name.
* `vt` (Number): *optional* * The length of time, in seconds, that a message received from a queue will be invisible to other receiving components when they ask to receive messages. Allowed values: 0-9999999 (around 115 days)
* `delay` (Number): *optional* The time in seconds that the delivery of all new messages in the queue will be delayed. Allowed values: 0-9999999 (around 115 days)
* `maxsize` (Number): *optional* The maximum message size in bytes. Allowed values: 1024-65536 and -1 (for unlimited size)

Note: At least one attribute (vt, delay, maxsize) must be supplied. Only attributes that are supplied will be modified.

Returns an object:

* `vt` (Number): The visibility timeout for the queue in seconds
* `delay` (Number): The delay for new messages in seconds
* `maxsize` (Number): The maximum size of a message in bytes
* `totalrecv` (Number): Total number of messages received from the queue
* `totalsent` (Number): Total number of messages sent to the queue
* `created` (Number): Timestamp (epoch in seconds) when the queue was created
* `modified` (Number): Timestamp (epoch in seconds) when the queue was last modified with `setQueueAttributes`
* `msgs` (Number): Current number of messages in the queue
* `hiddenmsgs` (Number): Current number of hidden / not visible messages. A message can be hidden while "in flight" due to a `vt` parameter or when sent with a `delay`

Example:

```javascript
rsmq.setQueueAttributes({ qname: "myqueue", vt: "30"}, function (err, resp) {
	if (err) {
		console.error(err)
		return
	}

	console.log("changed the invisibility time of messages that have been received to 30 seconds");
	console.log(resp);
});
```

### Messages

#### sendMessage
Sends a new message.

Parameters:

* `qname` (String)
* `message` (String)
* `delay` (Number): *optional* *(Default: queue settings)* The time in seconds that the delivery of the message will be delayed. Allowed values: 0-9999999 (around 115 days)
* `id` (String): *optional* *(Default: auto-generated)* Must be 32 characters from the alphabet `{A-Z, a-z, 0-9, :}` 

Returns:

* `id` (String): The internal message id.

Example:

```javascript
rsmq.sendMessage({ qname: "myqueue", message: "Hello World "}, function (err, resp) {
	if (err) {
		console.error(err)
		return
	}

	console.log("Message sent. ID:", resp);
});
```

#### receiveMessage(options, callback)
Receive the next message from the queue.

Parameters:

* `qname` (String): The Queue name.
* `vt` (Number): *optional* *(Default: queue settings)* The length of time, in seconds, that the received message will be invisible to others. Allowed values: 0-9999999 (around 115 days)

Returns an object:

  * `message` (String): The message's contents.
  * `id` (String): The internal message id.
  * `sent` (Number): Timestamp of when this message was sent / created.
  * `fr` (Number): Timestamp of when this message was first received.
  * `rc` (Number): Number of times this message was received.

Note: Will return an empty object if no message is there  

Example:

```javascript

CHANGELOG.md
# CHANGELOG for rsmq

## 0.12.4

* Added Travis tests for Node 16
* Updated dependencies
* Removed Travis tests for Node 10
## 0.12.3

* Added .NET implementation to README.md (#126)
* Fixed index.d.ts file to #110. (#123)
* Added rsmq-promise-native to README.md (#122)
* Removed support for Node 8
* Updated dependencies
## 0.12.2

* Added callback to `quit` method (#110)

## 0.12.1

* Added Travis tests for Node 14

## 0.12.0

* Removed Travis tests for Node 6
* Updated dependencies
* Added README for RSMQ Rust port. Thanks @DavidBM. (#107)

## 0.11.0

* Refactored all code from coffeescript to typescript.
* Added tests for #90
* Removed support for Node 4
* Support for `password` option when connecting to Redis (#70)

## 0.10.0

* added optional Promise support and tests

## 0.9.3

* removed `hiredis` optionalDependency

## 0.9.2

* add realtime option in typescript typings

## 0.9.0

* Added realtime option for Redis PUB/SUB when a new message is sent to RSMQ
* Added tests for realtime
* Added Travis tests for Node 10
* Updated dependencies (lodash and dev.async)

## 0.8.4

* Added Typescript typings

## 0.8.3

* Removed Travis test for Node 7
* Added Travis test for Node 8
* Mocha 4.x fixes for Travis test runner.

## 0.8.2

* Added mention for "RSMQ in other languages" in README.md

## 0.8.1

* Make sure `setQueueAttributes` does refresh `modified` field. Fix for #47. Thanks @igr

## 0.8.0

* Allow unlimited message size with `maxsize=-1` option.

## 0.7.2

* Removed Travis tests for Node 0.12.x

## 0.7.1

* Fix #30 Increased queue name limit from 80 to 160 chars.
* Run Travis tests latest Node 6.x, 4.x

## 0.7.0 

* Updated dependencies

## 0.6.0

* Added `popMessage` method
* Fix #23: Use of external Redis instance
* Added Tests for `popMessage`
* Use current version of lodash (4.5.1) and redis (2.4.2)

## 0.4.0

* Updated `redis` / `hiredis` modules.
* Node 0.8.x is no longer supported.
* Removed Travis tests for iojs
* Travis tests for Node 4.1 and 5.0

## 0.3.16

* Docs (Redis 2.6+ version requirement)

## 0.3.15

* Added LICENSE.md 
* Docs (added Links to modules)

## 0.3.14

* Fix `changeMessageVisibility` syntax fix. Failed if this method will be called as first call.

## 0.3.13

* Fix `hiddenmsgs` display in `getQueueAttributes`

## 0.3.12

* Added `quit` method

## 0.3.11

* Docs

## 0.3.10

* Docs

## 0.3.9 

* Added logo to README.md

## 0.3.8

* implemented `setQueueAttributes`
* switched from underscore to lodash
* added Travis test for Node.js 0.11
* updated the docs
* added tests for `setQueueAttributes`

## 0.3.5

* Make `hiredis` optional.

## 0.3.4

* Added support for [https://github.com/mranney/node_redis#rediscreateclientport-host-options](redis.createClient) `options` object.

## 0.3.3

* docs

## 0.3.2

* Added constructor option `client` to reuse existing redis clients

## 0.3.1

* Add details to README.md for constructor object. #3



index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("lodash");
const RedisInst = require("redis");
const EventEmitter = require("events").EventEmitter;
class RedisSMQ extends EventEmitter {
    constructor(options = {}) {
        super(options);
        this.asyncify = (methodKey) => {
            const asyncMethodKey = methodKey + "Async";
            this[asyncMethodKey] = (...args) => {
                return new Promise((resolve, reject) => {
                    this[methodKey](...args, (err, result) => {
                        if (err) {
                            reject(err);
                            return;
                        }
                        resolve(result);
                    });
                });
            };
        };
        this.quit = (cb) => {
            if (cb === undefined) {
                cb = () => { };
            }
            this.redis.quit(cb);
        };
        this._getQueue = (qname, uid, cb) => {
            const mc = [
                ["hmget", `${this.redisns}${qname}:Q`, "vt", "delay", "maxsize"],
                ["time"]
            ];
            this.redis.multi(mc).exec((err, resp) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                if (resp[0][0] === null || resp[0][1] === null || resp[0][2] === null) {
                    this._handleError(cb, "queueNotFound");
                    return;
                }
                const ms = this._formatZeroPad(Number(resp[1][1]), 6);
                const ts = Number(resp[1][0] + ms.toString(10).slice(0, 3));
                const q = {
                    vt: parseInt(resp[0][0], 10),
                    delay: parseInt(resp[0][1], 10),
                    maxsize: parseInt(resp[0][2], 10),
                    ts: ts
                };
                if (uid) {
                    uid = this._makeid(22);
                    q.uid = Number(resp[1][0] + ms).toString(36) + uid;
                }
                cb(null, q);
            });
        };
        this.changeMessageVisibility = (options, cb) => {
            if (this._validate(options, ["qname", "id", "vt"], cb) === false)
                return;
            this._getQueue(options.qname, false, (err, q) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                if (this.changeMessageVisibility_sha1) {
                    this._changeMessageVisibility(options, q, cb);
                    return;
                }
                this.on("scriptload:changeMessageVisibility", () => {
                    this._changeMessageVisibility(options, q, cb);
                });
            });
        };
        this._changeMessageVisibility = (options, q, cb) => {
            this.redis.evalsha(this.changeMessageVisibility_sha1, 3, `${this.redisns}${options.qname}`, options.id, q.ts + options.vt * 1000, (err, resp) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                cb(null, resp);
            });
        };
        this.createQueue = (options, cb) => {
            const key = `${this.redisns}${options.qname}:Q`;
            options.vt = options.vt != null ? options.vt : 30;
            options.delay = options.delay != null ? options.delay : 0;
            options.maxsize = options.maxsize != null ? options.maxsize : 65536;
            if (this._validate(options, ["qname", "vt", "delay", "maxsize"], cb) === false)
                return;
            this.redis.time((err, resp) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                const mc = [
                    ["hsetnx", key, "vt", options.vt],
                    ["hsetnx", key, "delay", options.delay],
                    ["hsetnx", key, "maxsize", options.maxsize],
                    ["hsetnx", key, "created", resp[0]],
                    ["hsetnx", key, "modified", resp[0]],
                ];
                this.redis.multi(mc).exec((err, resp) => {
                    if (err) {
                        this._handleError(cb, err);
                        return;
                    }
                    if (resp[0] === 0) {
                        this._handleError(cb, "queueExists");
                        return;
                    }
                    this.redis.sadd(`${this.redisns}QUEUES`, options.qname, (err, resp) => {
                        if (err) {
                            this._handleError(cb, err);
                            return;
                        }
                        cb(null, 1);
                    });
                });
            });
        };
        this.deleteMessage = (options, cb) => {
            if (this._validate(options, ["qname", "id"], cb) === false)
                return;
            const key = `${this.redisns}${options.qname}`;
            const mc = [
                ["zrem", key, options.id],
                ["hdel", `${key}:Q`, `${options.id}`, `${options.id}:rc`, `${options.id}:fr`]
            ];
            this.redis.multi(mc).exec((err, resp) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                if (resp[0] === 1 && resp[1] > 0) {
                    cb(null, 1);
                }
                else {
                    cb(null, 0);
                }
            });
        };
        this.deleteQueue = (options, cb) => {
            if (this._validate(options, ["qname"], cb) === false)
                return;
            const key = `${this.redisns}${options.qname}`;
            const mc = [
                ["del", `${key}:Q`, key],
                ["srem", `${this.redisns}QUEUES`, options.qname]
            ];
            this.redis.multi(mc).exec((err, resp) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                if (resp[0] === 0) {
                    this._handleError(cb, "queueNotFound");
                    return;
                }
                cb(null, 1);
            });
        };
        this.getQueueAttributes = (options, cb) => {
            if (this._validate(options, ["qname"], cb) === false)
                return;
            const key = `${this.redisns}${options.qname}`;
            this.redis.time((err, resp) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                const mc = [
                    ["hmget", `${key}:Q`, "vt", "delay", "maxsize", "totalrecv", "totalsent", "created", "modified"],
                    ["zcard", key],
                    ["zcount", key, resp[0] + "000", "+inf"]
                ];
                this.redis.multi(mc).exec((err, resp) => {
                    if (err) {
                        this._handleError(cb, err);
                        return;
                    }
                    if (resp[0][0] === null) {
                        this._handleError(cb, "queueNotFound");
                        return;
                    }
                    const o = {
                        vt: parseInt(resp[0][0], 10),
                        delay: parseInt(resp[0][1], 10),
                        maxsize: parseInt(resp[0][2], 10),
                        totalrecv: parseInt(resp[0][3], 10) || 0,
                        totalsent: parseInt(resp[0][4], 10) || 0,
                        created: parseInt(resp[0][5], 10),
                        modified: parseInt(resp[0][6], 10),
                        msgs: resp[1],
                        hiddenmsgs: resp[2]
                    };
                    cb(null, o);
                });
            });
        };
        this._handleReceivedMessage = (cb) => {
            return (err, resp) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                if (!resp.length) {
                    cb(null, {});
                    return;
                }
                const o = {
                    id: resp[0],
                    message: resp[1],
                    rc: resp[2],
                    fr: Number(resp[3]),
                    sent: Number(parseInt(resp[0].slice(0, 10), 36) / 1000)
                };
                cb(null, o);
            };
        };
        this.initScript = () => {
            const script_popMessage = `local msg = redis.call("ZRANGEBYSCORE", KEYS[1], "-inf", KEYS[2], "LIMIT", "0", "1")
			if #msg == 0 then
				return {}
			end
			redis.call("HINCRBY", KEYS[1] .. ":Q", "totalrecv", 1)
			local mbody = redis.call("HGET", KEYS[1] .. ":Q", msg[1])
			local rc = redis.call("HINCRBY", KEYS[1] .. ":Q", msg[1] .. ":rc", 1)
			local o = {msg[1], mbody, rc}
			if rc==1 then
				table.insert(o, KEYS[2])
			else
				local fr = redis.call("HGET", KEYS[1] .. ":Q", msg[1] .. ":fr")
				table.insert(o, fr)
			end
			redis.call("ZREM", KEYS[1], msg[1])
			redis.call("HDEL", KEYS[1] .. ":Q", msg[1], msg[1] .. ":rc", msg[1] .. ":fr")
			return o`;
            const script_receiveMessage = `local msg = redis.call("ZRANGEBYSCORE", KEYS[1], "-inf", KEYS[2], "LIMIT", "0", "1")
			if #msg == 0 then
				return {}
			end
			redis.call("ZADD", KEYS[1], KEYS[3], msg[1])
			redis.call("HINCRBY", KEYS[1] .. ":Q", "totalrecv", 1)
			local mbody = redis.call("HGET", KEYS[1] .. ":Q", msg[1])
			local rc = redis.call("HINCRBY", KEYS[1] .. ":Q", msg[1] .. ":rc", 1)
			local o = {msg[1], mbody, rc}
			if rc==1 then
				redis.call("HSET", KEYS[1] .. ":Q", msg[1] .. ":fr", KEYS[2])
				table.insert(o, KEYS[2])
			else
				local fr = redis.call("HGET", KEYS[1] .. ":Q", msg[1] .. ":fr")
				table.insert(o, fr)
			end
			return o`;
            const script_changeMessageVisibility = `local msg = redis.call("ZSCORE", KEYS[1], KEYS[2])
			if not msg then
				return 0
			end
			redis.call("ZADD", KEYS[1], KEYS[3], KEYS[2])
			return 1`;
            this.redis.script("load", script_popMessage, (err, resp) => {
                if (err) {
                    console.log(err);
                    return;
                }
                this.popMessage_sha1 = resp;
                this.emit("scriptload:popMessage");
            });
            this.redis.script("load", script_receiveMessage, (err, resp) => {
                if (err) {
                    console.log(err);
                    return;
                }
                this.receiveMessage_sha1 = resp;
                this.emit("scriptload:receiveMessage");
            });
            this.redis.script("load", script_changeMessageVisibility, (err, resp) => {
                if (err) {
                    console.log(err);
                    return;
                }
                this.changeMessageVisibility_sha1 = resp;
                this.emit('scriptload:changeMessageVisibility');
            });
        };
        this.listQueues = (cb) => {
            this.redis.smembers(`${this.redisns}QUEUES`, (err, resp) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                cb(null, resp);
            });
        };
        this.popMessage = (options, cb) => {
            if (this._validate(options, ["qname"], cb) === false)
                return;
            this._getQueue(options.qname, false, (err, q) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                if (this.popMessage_sha1) {
                    this._popMessage(options, q, cb);
                    return;
                }
                this.on("scriptload:popMessage", () => {
                    this._popMessage(options, q, cb);
                });
            });
        };
        this.receiveMessage = (options, cb) => {
            if (this._validate(options, ["qname"], cb) === false)
                return;
            this._getQueue(options.qname, false, (err, q) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                options.vt = options.vt != null ? options.vt : q.vt;
                if (this._validate(options, ["vt"], cb) === false)
                    return;
                if (this.receiveMessage_sha1) {
                    this._receiveMessage(options, q, cb);
                    return;
                }
                this.on("scriptload:receiveMessage", () => {
                    this._receiveMessage(options, q, cb);
                });
            });
        };
        this._popMessage = (options, q, cb) => {
            this.redis.evalsha(this.popMessage_sha1, 2, `${this.redisns}${options.qname}`, q.ts, this._handleReceivedMessage(cb));
        };
        this._receiveMessage = (options, q, cb) => {
            this.redis.evalsha(this.receiveMessage_sha1, 3, `${this.redisns}${options.qname}`, q.ts, q.ts + options.vt * 1000, this._handleReceivedMessage(cb));
        };
        this.sendMessage = (options, cb) => {
            if (this._validate(options, ["qname"], cb) === false)
                return;
            this._getQueue(options.qname, options["id"] == null, (err, q) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                options.delay = options.delay != null ? options.delay : q.delay;
                options.id = options.id != null ? options.id : q.uid;
                if (this._validate(options, ["delay"], cb) === false)
                    return;
                if (typeof options.message !== "string") {
                    this._handleError(cb, "messageNotString");
                    return;
                }
                if (q.maxsize !== -1 && options.message.length > q.maxsize) {
                    this._handleError(cb, "messageTooLong");
                    return;
                }
                const key = `${this.redisns}${options.qname}`;
                const mc = [
                    ["zadd", key, q.ts + options.delay * 1000, options.id],
                    ["hset", `${key}:Q`, options.id, options.message],
                    ["hincrby", `${key}:Q`, "totalsent", 1]
                ];
                if (this.realtime) {
                    mc.push(["zcard", key]);
                }
                this.redis.multi(mc).exec((err, resp) => {
                    if (err) {
                        this._handleError(cb, err);
                        return;
                    }
                    if (this.realtime) {
                        this.redis.publish(`${this.redisns}rt:${options.qname}`, resp[3]);
                    }
                    cb(null, options.id);
                });
            });
        };
        this.setQueueAttributes = (options, cb) => {
            const props = ["vt", "maxsize", "delay"];
            let k = [];
            for (let item of props) {
                if (options[item] != null) {
                    k.push(item);
                }
            }
            if (k.length === 0) {
                this._handleError(cb, "noAttributeSupplied");
                return;
            }
            if (this._validate(options, ["qname"].concat(k), cb) === false)
                return;
            const key = `${this.redisns}${options.qname}`;
            this._getQueue(options.qname, false, (err, q) => {
                if (err) {
                    this._handleError(cb, err);
                    return;
                }
                this.redis.time((err, resp) => {
                    if (err) {
                        this._handleError(cb, err);
                        return;
                    }
                    const mc = [
                        ["hset", `${this.redisns}${options.qname}:Q`, "modified", resp[0]]
                    ];
                    for (let item of k) {
                        mc.push(["hset", `${this.redisns}${options.qname}:Q`, item, options[item]]);
                    }
                    ;
                    this.redis.multi(mc).exec((err) => {
                        if (err) {
                            this._handleError(cb, err);
                            return;
                        }
                        this.getQueueAttributes(options, cb);
                    });
                });
            });
        };
        this._handleError = (cb, err, data = {}) => {
            let _err = null;
            if (_.isString(err)) {
                _err = new Error();
                _err.name = err;
                let ref = null;
                _err.message = ((ref = this._ERRORS) != null ? typeof ref[err] === "function" ? ref[err](data) : void 0 : void 0) || "unkown";
            }
            else {
                _err = err;
            }
            cb(_err);
        };
        this._initErrors = () => {
            this._ERRORS = {};
            for (let key in this.ERRORS) {
                this._ERRORS[key] = _.template(this.ERRORS[key]);
            }
        };
        this._VALID = {
            qname: /^([a-zA-Z0-9_-]){1,160}$/,
            id: /^([a-zA-Z0-9:]){32}$/
        };
        this._validate = (o, items, cb) => {
            for (let item of items) {
                switch (item) {
                    case "qname":
                    case "id":
                        if (!o[item]) {
                            this._handleError(cb, "missingParameter", { item: item });
                            return false;
                        }
                        o[item] = o[item].toString();
                        if (!this._VALID[item].test(o[item])) {
                            this._handleError(cb, "invalidFormat", { item: item });
                            return false;
                        }
                        break;
                    case "vt":
                    case "delay":
                        o[item] = parseInt(o[item], 10);
                        if (_.isNaN(o[item]) || !_.isNumber(o[item]) || o[item] < 0 || o[item] > 9999999) {
                            this._handleError(cb, "invalidValue", { item: item, min: 0, max: 9999999 });
                            return false;
                        }
                        break;
                    case "maxsize":
                        o[item] = parseInt(o[item], 10);
                        if (_.isNaN(o[item]) || !_.isNumber(o[item]) || o[item] < 1024 || o[item] > 65536) {
                            if (o[item] !== -1) {
                                this._handleError(cb, "invalidValue", { item: item, min: 1024, max: 65536 });
                                return false;
                            }
                        }
                        break;
                }
            }
            ;
            return o;
        };
        this.ERRORS = {
            "noAttributeSupplied": "No attribute was supplied",
            "missingParameter": "No <%= item %> supplied",
            "invalidFormat": "Invalid <%= item %> format",
            "invalidValue": "<%= item %> must be between <%= min %> and <%= max %>",
            "messageNotString": "Message must be a string",
            "messageTooLong": "Message too long",
            "queueNotFound": "Queue not found",
            "queueExists": "Queue exists"
        };
        if (Promise) {
            _.forEach([
                "changeMessageVisibility",
                "createQueue",
                "deleteMessage",
                "deleteQueue",
                "getQueueAttributes",
                "listQueues",
                "popMessage",
                "receiveMessage",
                "sendMessage",
                "setQueueAttributes",
                "quit"
            ], this.asyncify);
        }
        const opts = _.extend({
            host: "127.0.0.1",
            port: 6379,
            options: {
                password: options.password || null
            },
            client: null,
            ns: "rsmq",
            realtime: false
        }, options);
        opts.options.host = opts.host;
        opts.options.port = opts.port;
        this.realtime = opts.realtime;
        this.redisns = opts.ns + ":";
        if (opts.client && options.client.constructor.name === "RedisClient") {
            this.redis = opts.client;
        }
        else {
            this.redis = RedisInst.createClient(opts);
        }
        this.connected = this.redis.connected || false;
        if (this.connected) {
            this.emit("connect");
            this.initScript();
        }
        this.redis.on("connect", () => {
            this.connected = true;
            this.emit("connect");
            this.initScript();
        });
        this.redis.on("error", (err) => {
            if (err.message.indexOf("ECONNREFUSED")) {
                this.connected = false;
                this.emit("disconnect");
            }
            else {
                console.error("Redis ERROR", err);
                this.emit("error");
            }
        });
        this._initErrors();
    }
    _formatZeroPad(num, count) {
        return ((Math.pow(10, count) + num) + "").substr(1);
    }
    _makeid(len) {
        let text = "";
        const possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
        let i = 0;
        for (i = 0; i < len; i++) {
            text += possible.charAt(Math.floor(Math.random() * possible.length));
        }
        return text;
    }
}
module.exports = RedisSMQ;
